### ISSUE TITLE
Convert Simple Structured Answer Goodies to Full Templates

### ISSUE THREAD

**Issue by moollaza** [duckduckgo/zeroclickinfo-goodies#2767]
### Background

Goodies have full support for templates, and we'd like to make sure all Goodies are using our full templates. We provided the "simple structured_answer" as a way to get them using basic templates before they had full template support and now the time has come to upgrade them all.

### What To Do

1. Change the returned `structured_answer` object (in one of the .pm files below) to match the new format, which specifies a template and the data
    
    Nearly all of these Goodies should be able to use our simple `text` template and should return a `structured_answer` that looks like the one, returned by the PigLatin Goodie in [PigLatin.pm](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/lib/DDG/Goodie/PigLatin.pm#L40).

2. Update the Instant Answer's test file to match the changes made. The tests verify the expected content of the `structured_answer` for a given query. 

3. Test the IA locally with DuckPAN to make sure it still looks good and the tests pass. 

    You can run the test file with `duckpan test`:
    Example:
    Anagram.pm => `duckpan test Anagram`
    AspectRatio.pm => `duckpan test AspectRatio`

    You can test the visual display using `duckpan server`. You'll want to try some of the queries in the test file and make sure the visuals look good. Ideally identical to how they currently look on the live site.
    
### Example

@GuiltyDolphin has made [a great Pull Request](https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2769/files) that demonstrates the necessary changes for the Anagram Instant Answer:

<img width="999" alt="anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies" src="https://cloud.githubusercontent.com/assets/8598426/14021490/a110363e-f1d3-11e5-938f-faf11bef7016.png">

When updating the test file, it will make things _**much**_ easier if you create a small function that generates a structured answer object. This way you won't need to manually type out the full structured answer we expect for each individual query.

@GuiltyDolphin has done this in the Anagram PR as well:
<img width="918" alt="anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies" src="https://cloud.githubusercontent.com/assets/8598426/14021562/e795ccc2-f1d3-11e5-8817-cd16fe0531d7.png">

<img width="914" alt="anagrams__update_to_full_template_by_guiltydolphin_ _pull_request__2769_ _duckduckgo_zeroclickinfo-goodies" src="https://cloud.githubusercontent.com/assets/8598426/14021569/ed9a92ba-f1d3-11e5-9704-c0e5639e2712.png">

#### Note

Some of these IAs may actually be offline right now and so you won't be able to trigger them on DuckDuckGo.com. Once you've made a Pull Request we will deploy it to beta.duckduckgo.com so you can view your changes in a production environment

### How To Get Started

1. Find an Instant Answer in the list below that doesn't have a name beside it.
2. Leave a comment below that indicates which IA(s) you'd like to tackle. One of the community leaders or staff will put your name beside the IA in the list below indicating that it's been claimed.
3. Get coding and make a Pull Request.
    Be sure to include the link to the related Instant Answer Page in your PR Comment. This is **required** for us to be able to deploy your PR to Beta and ultimately merge it. For example, the anagram PR links to https://duck.co/ia/view/anagram.

    Also, make sure you mention this issue in your PR. That way we can keep track of all the PR's in here as well. Something like "Updates Anagram Goodie Re: #2767" is good.

----

Here's the list of IA's that need to be converted. If you'd like to work on one, please leave a comment letting everyone know, so we can avoid and duplicated efforts.

- [x] Anagram.pm        - @GuiltyDolphin
- [x] AspectRatio.pm    - @Mylainos
- [x] Base.pm    - @Mylainos
- [x] Average.pm   - @Mylainos
- [x] Base64.pm     - @Mylainos
- [x] BeamMeUpScotty.pm
- [x] Binary.pm   - @Mylainos
- [x] BirthStone.pm - @rasikapohankar
- [x] Bitsum.pm   - @Mylainos
- [x] Braille.pm -  https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2811
- [x] BRT.pm   - @Mylainos
- [x] CalcRoots.pm       - @GuiltyDolphin
- [x] CalendarConversion.pm   - @Mylainos
- [x] Calculator.pm        - @GuiltyDolphin
- [x] CallingCodes.pm   - @Mylainos
- [x] Chars.pm
- [x] CoffeeToWaterRatio.pm - @lamanh
- [x] Coin.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3128
- [x] Combination.pm
- [x] Constants.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3051
- [x] ConvertLatLon.pm - @mintsoft
- [x] CountryCodes.pm - @NaveenKarippai
- [x] CurrencyIn.pm
- [x] DaysBetween.pm - @NaveenKarippai https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3117
- [x] Factors.pm - @mintsoft
- [x] EmailValidator.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3163
- [x] Fibonacci.pm - @mintsoft
- [x] FlipText.pm - @NaveenKarippai
- [x] Fortune.pm - @mintsoft
- [x] GenerateMAC.pm - @mintsoft
- [x] GreatestCommonFactor.pm
- [x] GUID.pm - @gaulrobe #3007
- [x] HelpLine.pm - @mintsoft
- [x] HexToASCII.pm - @dbech
- [x] HexToDec.pm - @mintsoft
- [x] HTMLEntitiesDecode.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3041
- [x] ISO639.pm - @mintsoft
- [x] Jira.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3069
- [x] Kana.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3072
- [x] LeetSpeak.pm - @mintsoft
- [x] Lowercase.pm
- [x] MacAddress.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3076
- [x] MagicEightBall.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3077
- [x] MakeMeASandwich.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3078
- [x] MD4.pm - @mintsoft
- [x] MD5.pm - @mintsoft
- [x] MoonPhases.pm
- [x] Morse.pm
- [x] NoteFrequency.pm
- [x] PaleoIngredientCheck.pm
- [x] Paper.pm - @mintsoft
- [x] ParseCron.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3125
- [x] Passphrase.pm  - @mintsoft
- [x] Password.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3132
- [x] PercentOf.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3131
- [x] PhoneAlphabet.pm - @mintsoft
- [x] Pi.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3130
- [x] POTUS.pm
- [x] Rafl.pm - @mintsoft
- [x] RandomNumber.pm - @NaveenKarippai
- [x] Rc4.pm- @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3098
- [x] Reverse.pm - @mintsoft
- [x] ReverseComplement.pm - @mintsoft
- [x] RIPEMD.pm - @mintsoft
- [x] Roman.pm - @mintsoft
- [x] Rot13.pm - @mintsoft
- [x] Scramble.pm - @mintsoft
- [x] Sha.pm - @gaulrobe https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3119
- [x] SHA3.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3122
- [X] SolarSystem.pm
- [x] Sort.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3123
- [x] BeamMeUpScotty.pm
- [x] TimezoneConverter.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3124
- [x] TitleCase.pm
- [x] TwelveOclock.pm - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3127
- [x] Uppercase.pm
- [x] Wavelength.pm - @mintsoft
- [x] Week.pm  - @mintsoft https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3126
- [x] WeekdaysBetween.pm - https://github.com/duckduckgo/zeroclickinfo-goodies/pull/3104
- [x] Weight.pm
- [x] ValarMorghulis.pm
- [x] WhereAmI.pm - https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2684
- [x] WorkdaysBetween.pm
- [x] ZappBrannigan.pm - @rasikapohankar

---

/cc @duckduckgo/community-leaders

**Comment by GuiltyDolphin** [comment]
@moollaza Calculator is done in #1927 :wink: I'll tick it off.

**Comment by GuiltyDolphin** [comment]
@duckduckgo/community-leaders I'd just like to add something to this (optional). During the conversion, make a function for creating the template in the tests. It is so difficult to wade through tests that define the structured answer ***every_single_time***!

**Comment by GuiltyDolphin** [comment]
@moollaza I'll do Anagrams, seeing as it's first.

**Comment by moollaza** [comment]
> @moollaza I'll do Anagram, seeing as it's first.

@GuiltyDolphin Thanks!

**Comment by GuiltyDolphin** [comment]
@moollaza I'll do CalcRoots too.

**Comment by GuiltyDolphin** [comment]
Oh, boy, this could be interesting :wink:

For all yea in @duckduckgo/duckduckhack-contributors, we've got a lot of these to get through. Anyone interested in taking one up?

**Comment by faustow** [comment]
Me! Where/how can I get started?

**Comment by w2pc** [comment]
I have lots of free time at the office. Can I can take a few if I can get a handle on the first one.

**Comment by GuiltyDolphin** [comment]
@faustow @w2pc Heey!

Make sure you've got a passing familiarity with [the docs](http://docs.duckduckhack.com/) first, make sure you check out the [section on templates](http://docs.duckduckhack.com/frontend-reference/templates-overview.html) (that is what we are converting *to*).

If you look at #2769 and #2775, those show the general conversion we want to make (conversion to template, and update tests to easier to maintain format if possible).

If you get stuck at all just ping me on Slack or in your PR.

**Comment by moollaza** [comment]
> I have lots of free time at the office. Can I can take a few if I can get a handle on the first one.

@w2pc that'd be great! Thanks for helping out. Ping us on Slack if you need help :+1:

**Comment by moollaza** [comment]
@faustow I've updated the description with some detailed instructions, let us know if that helps!

**Comment by faustow** [comment]
It does! Thanks a lot @moollaza

**Comment by rasikapohankar** [comment]
I would like to work on BirthStone.pm .

**Comment by lamanh** [comment]
Hi,
I'm working on **EmailValidator.pm**, and found a limitation on **Goodie.pm** (/home/codio/perl5/perls/perl-5.18.2/lib/site_perl/5.18.2/DDG/Test/Goodie.pm line: 65).

Current test can only handle Regex/-ANY- pattern for **one dimension** structured_answer. But the new template is two dimension. So tests use Regex/-ANY- will always fail.

Should we update **Goodie.pm** to handle multi-dimensional structured_answer?

**Comment by GuiltyDolphin** [comment]
@lamanh There is a PR for this already duckduckgo/duckduckgo#172. Hopefully that will enable regex matching arbitrarily deep.

**Comment by Mylainos** [comment]
I would like to work on Aspect Ratio.

**Comment by Mylainos** [comment]
I would like to work on Base please.

**Comment by Mylainos** [comment]
I will pick Average, Binary and Bitsum please.

**Comment by Mylainos** [comment]
I will pick BRT, CalendarConversion and CallingCodes please.

**Comment by GuiltyDolphin** [comment]
@Mylainos Hehe, thanks for all this! I imagine you're getting a pretty good feel for the available templates and options now, huh?

**Comment by Mylainos** [comment]
@GuiltyDolphin Oh yes! Now I know a lot of templates like 'text' and ... all the options that come with... I think I master the templates.

**Comment by Mylainos** [comment]
I will wait for BRT because of #2863

**Comment by Mylainos** [comment]
I will take Chars, CoffeeToWaterRatio and Coin please :)

**Comment by Mylainos** [comment]
Apparently CoffeeToWaterRatio have been already done here #2838 by @lamanh.

**Comment by Mylainos** [comment]
So for using regex when testing templates one should use Test::Deep ? It don't work for me...

**Comment by dbech** [comment]
Hey, I've done HexToASCII.pm in the pull above

**Comment by dbech** [comment]
I'll take Coin.pm as well if that's cool.

**Comment by tagawa** [comment]
Thank you @dbech - that sounds great!

**Comment by NaveenKarippai** [comment]
I'm highly interested to work on FlipText.

**Comment by NaveenKarippai** [comment]
I'd like to pick CoffeeToWaterRatio

**Comment by NaveenKarippai** [comment]
hm, CoffeeToWaterRatio seems to be already done on open PR #2819.

[double-check]:  SolarSystem looks like to be already running a full-template.

Please assign CountryCodes, EmailValidator, and RandomNumber to me.

**Comment by rasikapohankar** [comment]
I'm working on ZappBrannigan.pm.
@NaveenKarippai I guess EmailValidator is being worked on, https://github.com/duckduckgo/zeroclickinfo-goodies/issues/2767#issuecomment-203883344

**Comment by NaveenKarippai** [comment]
@rasikapohankar Thanks for the heads up! I'm bumping into duplicates/already-taken tasks pretty often.

**Comment by GuiltyDolphin** [comment]
@NaveenKarippai Yup, looks like SolarSystem doesn't need updating; thanks for pointing that out!

**Comment by NaveenKarippai** [comment]
Please assign DaysBetween, Password, PercentOf and Pi to me.

**Comment by gaulrobe** [comment]
@moollaza #3119 fixes the Sha Goodie.

**Comment by mintsoft** [comment]
@NaveenKarippai You're currently assigned to: DaysBetween, Password.pm, PercentOf.pm,  Pi.pm

I can see that you're currently working on DaysBetween; do you mind if I unassign Password. PercentOf and Pi so someone else (probably myself) can quickly knock them out?

Thanks

**Comment by mintsoft** [comment]
@lamanh You're currently assigned to EmailValidator; I can see you started work on it already but hit some issues. Do you still want to be assigned to it?

**Comment by NaveenKarippai** [comment]
@mintsoft no worries! please go ahead.

**Comment by mintsoft** [comment]
@lamanh Just FYI; I've got a PR in progress for Email Validator: https://github.com/duckduckgo/zeroclickinfo-goodies/issues/3163

**Comment by GuiltyDolphin** [comment]
@moollaza @mintsoft Just two more to go! We need to have a partaay once this is done :tada: 

@mintsoft Thanks for doing so many of these - we'd still be a long way off if it weren't for you :heart:

Thanks to everyone else who's done one (or more) too; we really appreciate this community effort :dolphin:

**Comment by GuiltyDolphin** [comment]
@duckduckgo/community-leaders Time to partaay!

:tada: :balloon: :cake:
